libpretixui-android/src/main/java/eu/pretix/libpretixui/android/scanning/ScannerView.kt

```diff
@@ -1,30 +1,298 @@
-    override fun getFramingRectInPreview(previewWidth: Int, previewHeight: Int): Rect {
-        if (this.mFramingRectInPreview == null) {
-            val rect = Rect()
-            rect.left = 0
-            rect.top = 0
-            rect.right = previewWidth
-            rect.bottom = previewHeight
+    var torch: Boolean
+        get() = torchTarget
+        set(value) {
+            torchTarget = value
+            if (torchState != torchTarget) {
+                camera?.cameraControl?.enableTorch(torchTarget)
+            }
+        }
+
+    var autofocus: Boolean
+        get() = autofocusTarget
+        set(value) {
+            autofocusTarget = value
+            if (autofocusTarget != autofocusState) {
+                enableAutofocus(autofocusTarget)
+            }
+        }
+
+    override fun addView(child: View?) {
+        super.addView(child)
+    }
+
+    fun setResultHandler(rh: ResultHandler) {
+        this.resultHandler = rh
+    }
+
+    fun startCamera() {
+        removeAllViews()
+        previewView = PreviewView(context)
+        previewView!!.scaleType = PreviewView.ScaleType.FILL_CENTER
+        addView(previewView)
+
+        val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
+        cameraExecutor = Executors.newSingleThreadExecutor()
+
+        cameraProviderFuture.addListener({
+            bindPreview(cameraProviderFuture.get())
+        }, ContextCompat.getMainExecutor(context))
+    }
+
+    @androidx.camera.camera2.interop.ExperimentalCamera2Interop
+    private fun enableAutofocus(enabled: Boolean) {
+        if (camera == null) return
+        if (enabled) {
+            Camera2CameraControl.from(camera!!.cameraControl).captureRequestOptions =
+                CaptureRequestOptions.Builder()
+                    .setCaptureRequestOption(CaptureRequest.CONTROL_AF_MODE, CameraMetadata.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
+                    .build()
+        } else {
+            val chars = Camera2CameraInfo.from(camera!!.cameraInfo)
+            Camera2CameraControl.from(camera!!.cameraControl).captureRequestOptions =
+                CaptureRequestOptions.Builder()
+                    .setCaptureRequestOption(CaptureRequest.CONTROL_AF_MODE, CameraMetadata.CONTROL_AF_MODE_OFF)
+                    .setCaptureRequestOption(CaptureRequest.LENS_FOCUS_DISTANCE, chars.getCameraCharacteristic(CameraCharacteristics.LENS_INFO_MINIMUM_FOCUS_DISTANCE)!!)
+                    .build()
+        }
+        autofocusState = enabled
+    }
+
+    private fun bindPreview(cameraProvider: ProcessCameraProvider) {
+        this.cameraProvider = cameraProvider
+        cameraProvider.unbindAll()
+
+        val preview: Preview = Preview.Builder()
+            .build()
+        preview.setSurfaceProvider(previewView!!.surfaceProvider)
+
+        val cameraSelector: CameraSelector = CameraSelector.Builder()
+            .requireLensFacing(CameraSelector.LENS_FACING_BACK)
+            .build()
```

Comment:
Requiring the lens selector to use the back one breaks on some of my devices, so it results in no camera chosen. 
This relaxes that a bit:
```suggestion
        var cameraSelector: CameraSelector = CameraSelector.Builder().build()
        if (cameraProvider.hasCamera(CameraSelector.DEFAULT_BACK_CAMERA)) {
            cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA
        }
```